keel-slack/src/main/kotlin/com/netflix/spinnaker/keel/slack/NotificationEventListener.kt

```diff
@@ -56,30 +67,156 @@ class NotificationEventListener(
+      if (latestApprovedArtifactVersion == null) {
+        log.warn("latestApprovedArtifactVersion is null for application ${config.application}, env $targetEnvironment. Can't send UnpinnedNotification")
+        return
+      }
+
+      val latestArtifact = repository.getArtifactVersion(pinnedEnvironment!!.artifact, latestApprovedArtifactVersion, null)
+      val pinnedArtifact = repository.getArtifactVersion(pinnedEnvironment!!.artifact, pinnedEnvironment!!.version, null)
+
+      val handler = handlers.find {
+        it.type == notification.type
+      } as? SlackNotificationHandler<SlackUnpinnedNotification> ?: return
+
+      config.environments.forEach { environment ->
+        environment.notifications.filter {
+          it.type == NotificationType.slack
+        }.onEach {
+          handler.sendMessage(SlackUnpinnedNotification(
+            channel = it.address,
+            latestArtifact = latestArtifact,
+            pinnedArtifact = pinnedArtifact,
+            application = config.application,
+            time = clock.instant(),
+            user = user,
+            targetEnvironment = targetEnvironment
+          ))
+        }
+      }
+    }
+  }
+
+  @EventListener(MarkAsBadNotification::class)
+  fun onMarkAsBadNotification(notification: MarkAsBadNotification) {
+    with(notification) {
+      val deliveryArtifact = config.artifacts.find {
+        it.reference == veto.reference
+      } ?: return
+
+      val vetoedArtifact = repository.getArtifactVersion(deliveryArtifact, veto.version, null)
+      if (vetoedArtifact == null) {
+        log.warn("vetoedArtifact is null for application ${config.application}. Can't send MarkAsBadNotification notification")
+        return
+      }
+
+      val handler = handlers.find {
+        it.type == notification.type
+      } as? SlackNotificationHandler<SlackMarkAsBadNotification> ?: return
+
+      config.environments.forEach { environment ->
+        environment.notifications.filter {
+          it.type == NotificationType.slack
+        }.onEach {
+          handler.sendMessage(SlackMarkAsBadNotification(
+            vetoedArtifact = vetoedArtifact,
+            user = user,
+            targetEnvironment = veto.targetEnvironment,
+            channel = it.address,
+            time = clock.instant(),
+            application = config.name,
+            comment = veto.comment
+          ))
+        }
+      }
+    }
+  }
+
+  @EventListener(ApplicationActuationPaused::class)
+  fun onApplicationActuationPaused(notification: ApplicationActuationPaused) {
+    with(notification) {
+      val config = repository.getDeliveryConfigForApplication(application)
+
+      val handler = handlers.find {
+        it.type == Type.PAUSED_APPLICATION
+      } as? SlackNotificationHandler<SlackPausedNotification> ?: return
+
+      config.environments.forEach { environment ->
```

Comment:
Same question on looping over environments. BTW, this block of code repeats many places, so it's a good candidate to extract into a function. It could look something like:

```kotlin
fun Environment.sendSlackMessage(message: SlackNotificationEvent) {
  this.notifications.filter {
    it.type == NotificationType.slack
  }.forEach {
    handlers.supporting(message.type)?.sendMessage(message)
  }
```

Reply:
DONE! thank you!!